File: src/main/java/com/nanaios/polygonal_tech/capability/interfaces/IItemSlot.java
package com.nanaios.polygonal_tech.capability.interfaces;

import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.items.IItemHandlerModifiable;
import org.jetbrains.annotations.NotNull;

/// 単一のItemStackスロットを表すインターフェース
public interface IItemSlot extends IItemHandlerModifiable, ICapability {

    /// ItemStackを取得するメソッド
    ///
    /// @return スロットにあるItemStack。スロットが空の場合はItemStack.EMPTYを返す。
    @NotNull ItemStack getStack();

    /// ItemStackを設定するメソッド
    ///
    /// @param stack 設定するItemStack
    void setStack(@NotNull ItemStack stack);

    /// ItemStackを挿入するメソッド
    ///
    /// @param stack    挿入するItemStack
    /// @param simulate ItemStackを実際に挿入するかどうか。trueの場合、挿入できるかどうかをシミュレートするだけで、実際には挿入しません。
    /// @return 挿入できなかったItemStackの残り。
    @NotNull ItemStack insertItem(@NotNull ItemStack stack, boolean simulate);

    /// ItemStackを抽出するメソッド
    ///
    /// @param amount   抽出するItemStackの数量
    /// @param simulate ItemStackを実際に抽出するかどうか。trueの場合、抽出できるかどうかをシミュレートするだけで、実際には抽出しません。
    /// @return 抽出されたItemStack。スロットが空の場合はItemStack.EMPTYを返す。
    @NotNull ItemStack extractItem(int amount, boolean simulate);

    /// スロットの最大スタックサイズを返すメソッド
    int getSlotLimit();

    /// ItemStackがスロットに挿入可能かどうかを判断するメソッド
    ///
    /// @param stack 挿入しようとしているItemStack
    boolean isItemValid(@NotNull ItemStack stack);

    /// {@link net.minecraft.world.inventory.AbstractContainerMenu}などで、スロットのx座標を指定するためのメソッド
    int getMenuX();

    /// {@link net.minecraft.world.inventory.AbstractContainerMenu}などで、スロットのy座標を指定するためのメソッド
    int getMenuY();

    @Override
    default CompoundTag serializeNBT() {
        return getStack().serializeNBT();
    }

    @Override
    default void deserializeNBT(CompoundTag nbt) {
        ItemStack stack = ItemStack.of(nbt);
        if (!stack.isEmpty()) {
            setStack(stack);
            return;
        }

        setStack(ItemStack.EMPTY);
    }

    @Override
    default int getSlots() {
        return 1;
    }

    @Override
    @NotNull
    default ItemStack getStackInSlot(int slot) {
        return slot == 0 ? getStack() : ItemStack.EMPTY;
    }

    @Override
    default void setStackInSlot(int slot, @NotNull ItemStack stack) {
        if (slot == 0) setStack(stack);
    }

    @Override
    @NotNull
    default ItemStack insertItem(int slot, @NotNull ItemStack stack, boolean simulate) {
        return slot == 0 ? insertItem(stack, simulate) : stack;
    }

    @Override
    @NotNull
    default ItemStack extractItem(int slot, int amount, boolean simulate) {
        return slot == 0 ? extractItem(amount, simulate) : ItemStack.EMPTY;
    }

    @Override
    default int getSlotLimit(int slot) {
        return slot == 0 ? getSlotLimit() : 0;
    }

    @Override
    default boolean isItemValid(int slot, @NotNull ItemStack stack) {
        return slot == 0 && isItemValid(stack);
    }
}

File: src/main/kotlin/com/nanaios/polygonal_tech/core/capability/item/IItemSlot.kt
package com.nanaios.polygonal_tech.core.capability.item

import com.nanaios.polygonal_tech.core.capability.ICapability
import net.minecraft.world.item.ItemStack
import net.minecraftforge.items.IItemHandlerModifiable

/**
 * 独自のインベントリを持つ機械等において、GUI上の表示位置（[x], [y]）を関連付けた1スロット分のアイテム情報を保持し、
 * ネットワーク同期等の[ICapability]としての特性を持たせることを目的としたインターフェース。
 */
interface IItemSlot: IItemHandlerModifiable, ICapability {
    /** GUI上でこのスロットが配置されるべきX座標 */
    val x:Int
    /** GUI上でこのスロットが配置されるべきY座標 */
    val y:Int
    /** 現在このスロットに入っている[ItemStack] */
    var stack: ItemStack
    /** このスロットが保持できるアイテムの最大数 */
    val slotLimit: Int

    /**
     * スロットにアイテムを搬入することを目的とするメソッド。
     *
     * @param stack 搬入したい[ItemStack]
     * @param simulate 実際の搬入を行うか、シミュレートかを示すフラグ
     * @return 搬入しきれず余った[ItemStack]。全て入った場合は[ItemStack.EMPTY]を返す。
     */
    fun insertItem(stack: ItemStack, simulate: Boolean): ItemStack

    /**
     * スロットからアイテムを搬出することを目的とするメソッド。
     *
     * @param amount 搬出したい最大数量
     * @param simulate 実際の搬出を行うか、シミュレートかを示すフラグ
     * @return 実際に搬出された[ItemStack]
     */
    fun extractItem(amount: Int, simulate: Boolean): ItemStack

    /**
     * 対象のアイテムがこのスロットの条件を満たしており、搬入可能かを検証することを目的とするメソッド。
     *
     * @param stack 検証対象の[ItemStack]
     * @return 搬入可能であれば[true]
     */
    fun isItemValid(stack: ItemStack): Boolean

    override fun getSlots() = 1
    override fun getSlotLimit(slot: Int) = slotLimit
    override fun getStackInSlot(slot: Int) = stack
    override fun setStackInSlot(slot: Int, stack: ItemStack) { this.stack = stack }
    override fun insertItem(slot: Int, stack: ItemStack, simulate: Boolean) = insertItem(stack, simulate)
    override fun extractItem(slot: Int, amount: Int, simulate: Boolean) = extractItem(amount, simulate)
    override fun isItemValid(slot: Int, stack: ItemStack) = isItemValid(stack)
}